Add file class structure

kconnour opened the pull request:
This is a follow up to #19.

At long last here's the much ballyhooed class structure. I haven't included very many public methods but I really think this can be a skeleton from which we can add basic file handling methods. The idea is no "advanced" file routines go in this class but these classes could be input into other classes/functions for more advanced utilities. I can and plan to add several more methods, but I figured I may as well commit this code since it's useable as is.

I included a hierarchy of classes and occasional methods slotted into the proper classes. If you don't want to work with classes directly, I added 3 helper functions that essentially just create patterns that play well with the classes. These classes are inspired by Zac's request a while ago.

I encourage any and all input on the details of this implementation. While this commit breaks the rule of avoiding code duplication, recall we made an exception for this implementation... therefore I put the code in its own file to isolate it somewhat from the rest of the repo.

planetarymike commented:
Thanks Kyle! I'll take a look at this today.

kconnour commented:
You may already know this but there's a way to ensure a [certain number of people sign off on a pull request](https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/enabling-required-reviews-for-pull-requests) before it's merged. I get the feeling you and I care way more about this request than Zac so I didn't include him as a reviewer, but perhaps down the line we can loop him in as a reviewer. That'd be a good learning opportunity, at least for me :)

planetarymike commented:
Hmm, I don't have access to settings on this repo--- I'll see if Josh can add us.

planetarymike commented:
I enabled the require reviewers option for merge to master, now we need two reviews before code can be merged.

planetarymike reviewed (commented):
Thanks for pushing this, Kyle--- I learned a fair amount about Python objects. Coming from a C++ background, it's frustrating to me that class member variables aren't declared anywhere and can be introduced by any member function. Makes it harder to keep track of what's in state for the object. But I'll get used to it :)

Looking at the code as a whole, it seems to me that one big benefit is that the object returned has methods for getting absolute filenames as well as the basenames. This avoids having to keep track of two objects if both are needed. In general I'm in favor of implementing an object like this and extending its capabilities, but I have some concerns about this implementation.

The big one is that I'm not sure what the subclasses are doing other than implementing some checks and throwing errors that might be better implemented as changes in the glob expression being searched for. For example, the IUVSFiles class could prepend 'mvn_iuv\*' to the pattern the user requests rather than throwing an error when some files don't contain this pattern. same thing with '\*l1b\*' and "\*.fits\*". The relay check would not fit into this scheme, but since that involves examining file contents I'm pretty sure it should be handled by a subclass or function specific to relay files.

There are a few things that would be good to maintain from the various implementations of file finders we have now:

1. Speed. Globbing through the whole l1b dataset for a small number of files is noticeably slow. Zac and I solved this problem in different ways: Zac by figuring out what orbit directory to look in for an single orbit's worth of files, and myself by implementing an index file. Ideally we'd keep both methods with a `use_index` flag and a routine to parse the orbit number(s) requested.
2. Automatic location of l1b directory. If the user has set up the l1b directory in `user_paths.py`, the routine should interpret `path=None` as an instruction to use this directory. I have this as the default in `maven_iuvs.search.find_all_l1b` I think all of us have used defaults like this in the past, and I'd like to keep it so I don't need to type in my file path all the time.
3. User protection. This would implement some of the stuff in your class through insertion of tags such as "mvn_iuv\*", "\*l1b\*", and "\*.fits\*" before the glob call. Other protections such as the path existence check would need to be implemented elsewhere.
4. Returning an object with convenience functions to make the user's life easier. 

For the object returned by the finder routine, here are some things that would be nice:

- Properties defined in your classes, including .filenames, .filebasenames, and .count
- Your methods to downselect the filename based on a pattern, which should either modify the object contents in place or return an object with a subset of the files.
- Methods for `__str__` and `__repr__` set to print either the absolute or relative filenames in the style of a numpy array when the object is used in an interactive session
- Definitions of `__iter__()` and `__next__()` that users can use to transparently loop over the absolute file names to open the FITS files or do other things. 

I think the ultimate outcome here would be a single file finder function that combines all the features of Zac's get_files() and my find_all_l1b(), plus potentially some other option flags, and returns a file object with the properties listed above. 

Negotiable on all of the above, of course! If we can agree on a path forward, I'd be happy to write some of the code. Let me know what you think :)

kconnour commented:
Thanks for taking a look at this in depth. I'm certainly amenable to changes and will respond to your in-depth code comments after Christmas. As you can probably infer, I'm prioritizing organization and readability above all else. If I interpret you correctly, many of your comments revolve around useability and my thoughts are almost always "we can just add that in a helper function that creates one of these objects." So with that in mind, here are some of my thoughts.

> Thanks for pushing this, Kyle--- I learned a fair amount about Python objects. Coming from a C++ background, it's frustrating to me that class member variables aren't declared anywhere and can be introduced by any member function. Makes it harder to keep track of what's in state for the object. But I'll get used to it :)

What's a class member variable? It sounds like a class attribute/property, which I define with `@property`, so maybe I'm missing something. Fortunately (at least from my perspective), I made these classes as immutable as possible by including getters but no setters. 

> Looking at the code as a whole, it seems to me that one big benefit is that the object returned has methods for getting absolute filenames as well as the basenames. This avoids having to keep track of two objects if both are needed. In general I'm in favor of implementing an object like this and extending its capabilities, but I have some concerns about this implementation.

Aren't these just stored in the class' properties?

> The big one is that I'm not sure what the subclasses are doing other than implementing some checks and throwing errors that might be better implemented as changes in the glob expression being searched for. For example, the IUVSFiles class could prepend 'mvn_iuv*' to the pattern the user requests rather than throwing an error when some files don't contain this pattern. same thing with '*l1b*' and "*.fits*". The relay check would not fit into this scheme, but since that involves examining file contents I'm pretty sure it should be handled by a subclass or function specific to relay files.

Again, I've only included a few methods so a lot of the subclasses that don't do much are branch points. For example, I think when I clean up my database creation code, it could extend `IUVSFiles`---that way it knows the filenames and file paths, and has checks done to ensure it's getting an IUVS file as a starting point. Similarly, I expect someone may create an L1cFiles class that extends `IUVSDataFiles`. So while `IUVSFiles` and `IUVSDataFiles` each barely do anything, in the above example I think it'd be advantageous to have a structure like this for someone wanting to branch off these classes when adding their own utilities. I believe this also provides us an organizational structure too since we could just add methods to the appropriate class instead of just adding functions to a file. 

Regarding appending strings to the input, I think this is ultimately a design choice we'll have to make if we go with this class structure... but I would like to make an argument for the structure as it is. Since `Files` just takes `pattern` and `path` I find it much cleaner to have all subclasses in this scheme accept the exact same arguments---that way I know if I want to create any of the classes what the arguments should be. If we appended strings, I'd have to remember that, say, `IUVSDataFiles` is a subclass of `IUVSFiles` so I shouldn't include "mvn_iuv*" because that class already does that for me. In such a scheme I'd really have to know the hierarchy of class extensions.

However, I do like the idea of the user having to do less work and not having to append all the tags to the search pattern. Much like numpy discourages people from creating ndarrays direclty, I would propose that we make functions that simply append "mvn_iuv" or ".fits"  to create the `pattern` string. I think this combines the best of both worlds---the class inputs are still general and the same for all class, and the functions make it so no one needs to know the class structure and are user-friendly. 

> There are a few things that would be good to maintain from the various implementations of file finders we have now:
> 
>     1. Speed. Globbing through the whole l1b dataset for a small number of files is noticeably slow. Zac and I solved this problem in different ways: Zac by figuring out what orbit directory to look in for an single orbit's worth of files, and myself by implementing an index file. Ideally we'd keep both methods with a `use_index` flag and a routine to parse the orbit number(s) requested.

>     2. Automatic location of l1b directory. If the user has set up the l1b directory in `user_paths.py`, the routine should interpret `path=None` as an instruction to use this directory. I have this as the default in `maven_iuvs.search.find_all_l1b` I think all of us have used defaults like this in the past, and I'd like to keep it so I don't need to type in my file path all the time.

>     3. User protection. This would implement some of the stuff in your class through insertion of tags such as "mvn_iuv*", "*l1b*", and "*.fits*" before the glob call. Other protections such as the path existence check would need to be implemented elsewhere.

>     4. Returning an object with convenience functions to make the user's life easier.

I agree these would all be good things to have. I believe my classes have 1. built in (since I'm essentially doing "Zac's method" that you describe above) and I believe functions that return these classes could handle the other requests. 

> 
> For the object returned by the finder routine, here are some things that would be nice:
> 
>     * Properties defined in your classes, including .filenames, .filebasenames, and .count

I realized right after commiting that I could add n_files as a property so I'll do so. However I'm not really sure I understand what you're asking for regarding .filenames and .basenames since all classes here have both `filenames` and `absolute_file_paths` properties.

>     * Your methods to downselect the filename based on a pattern, which should either modify the object contents in place or return an object with a subset of the files.

Are you suggesting that I return a Files (or derived) object every time I call one of the methods in Files? I thought returning a list would be most convenient but you bring up an interesting suggestion.

>     * Methods for `__str__` and `__repr__` set to print either the absolute or relative filenames in the style of a numpy array when the object is used in an interactive session

Good idea. I've never done that before so it'll be something good to learn :)

>     * Definitions of `__iter__()`, `__next__()`, and the `__getitem__()` family that users can use to transparently loop over the absolute file names to open the FITS files or do other things.

Also a good idea. Do you have a preference if it loops over absolute file paths or just filenames?

 > I think the ultimate outcome here would be a single file finder function that combines all the features of Zac's get_files() and my find_all_l1b(), plus potentially some other option flags, and returns a file object with the properties listed above.

I'm not sure I like this idea but maybe you're thinking of something that I'm not. If you're on board with the class structures (at least generally), then there are times when you'd only ever want an `L1bFiles` object and other times when you'd want a `SingleOrbitModeSequenceL1bFiles` object. If we put all of the creation in one function, it'd return objects from different classes which I think could be potentially confusing. It'd also break the one responsibility principle for functions. 

> Negotiable on all of the above, of course! If we can agree on a path forward, I'd be happy to write some of the code. Let me know what you think :)

I imagine we can work something out! I've certainly argued my thoughts here but you may very well be thinking of arguments that I'm not. It may be most helpful to have a Zoom call sometime early next week. Perhaps sometime Monday afternoon? I'd happy to have your contributions but I think we should first decide on how we want the structure before we fill any of it in.

kconnour commented on maven_iuvs/files.py:
```diff
@@ -0,0 +1,444 @@
+
+        Notes
+        -----
+        This class used glob-style matching, so a pattern of ['**/*.pdf'] will recursively search for .pdf files
```
I'll fix that typo on the next commit. Thanks for pointing out PEP8... long story short there's a bug in Pycharm and it thinks PEP8's character limit is not 79 but 120, which I've been religiously adhering to. I'll try to configure it such that all my future commits adhere to that limit.

kconnour commented on maven_iuvs/files.py:
```diff
@@ -0,0 +1,444 @@
+
+    @property
+    def filenames(self):
+        """ Get filenames """
```
I found writing these felt awkward since I had documented those things in the constructor's docstring... which is usually a sign I'm doing something wrong. I moved info on all properties into a more verbose docstring associated with each property, which will be in the next commit. I'll mark this as resolved for now but open to suggestions on my improved way of documenting these.

kconnour commented on maven_iuvs/files.py:
```diff
@@ -0,0 +1,444 @@
+import numpy as np
+
+
+class Files:
```
I've never used pathlib but if it does all the work I'm doing, I may as well use it since it's probably more thoroughly tested.

kconnour replied:
I spent some time this afternoon looking at that library and so far I'm not impressed with some of the design choices they've made. I'll elect to keep `Files` as is but I'm amenable to changing the implementation to use this library in the future if someone's more competent with this library than I am.

kconnour commented on maven_iuvs/files.py:
```diff
@@ -0,0 +1,444 @@
+        path: str
+            The location where to start looking for files.
+        patterns: list
+            List of strings of regex patterns to match filenames to.
```
I didn't know that! I just changed all of them

kconnour commented on maven_iuvs/files.py:
```diff
@@ -0,0 +1,444 @@
+        try:
+            all_paths = []
+            for i in patterns:
+                all_paths.extend(glob.glob(os.path.join(path, i), recursive=True))
```
I'm open to suggestions! I kinda like this implementation because the `recursive=True` won't recursively search unless the user adds a **, but it allows that functionality if the user wants it. I do add a note about this functionality in the docstring so I'm not aware of anything that's overly hidden to the user.

kconnour commented:
Instead of responding to all of these, I pushed changes that address several of your comments. Still happy to chat about them sometime if you want.

planetarymike commented:
Let's hash out a general idea before either of us does more work to implement anything. I think there's a place for a filelist class, but I think we ought to keep the container and any file finding functions simple and easy to understand and maintain. 

I don't like the idea of inheritance here because inheritance and particularly multiple inheritance can lead to a lot of code maintenance problems down the line: contributors or maintainers would need to keep track of where to place new functionality (should a new method belong to l1bfiles or files? should a new routine be a class method at all, or live as a separate function?). It's likely to be a huge headache. 

Instead, I think we should have a base file list class that's totally agnostic to what sorts of files live inside and where they came from (accepting only a list of absolute file paths in the constructor), but adds some convenience functions that a generic array or ndarray wouldn't have. We could even subclass ndarray to incorporate its desirable features (though we definitely don't need all of its features). I can see a place for a basenames() method and a n_files method (or just use ndarray.size()). What else do we need? It seems like some of this should have been figured out and implemented already, maybe by Pathlib? What did you find out when you looked in more detail at the library?

For file finding, it seems to me that we should provide utilities to select based on the defined components of the filename, since those are well-defined for normal orbit data and unlikely to change. Putting all of the string constriction and glob functionality in a single function or handful of functions seems to be the cleanest from the point of view of understanding and maintaining the code. Right now with your classes the construction of the glob string is split up across multiple class constructors, which makes it hard to see what's being globbed for. If instead we had a function with arguments for each filename segment, this would put all of the functionality in one place. Using mvn_iuv_l1b_apoapse-orbit00236-muv_20141112T032125_v13_r02.fits.gz as an example, the function would have a signature like

`def find_files(path=None, level=None, segment=None, orbit=None, channel=None, time=None):`

This top level function would construct the search string, including specifying the appropriate orbit folders when an orbit range is passed, and return an `iuvs_filelist` object which has the convenience features needed by all files. I think we'd need a few more arguments for the function to capture the full featureset of what's available now between Zac's, yours, and my routines. 

People needing to define a function to return more specific sets of files based on inspecting the file contents could then write a function similar to Zac's `get_apoapse_files` incorporating a call to `find_files` and some file inspection routine that loops over an `iuvs_filelist` object and keeps only the correct files, along with some additional info. This would likely be way more transparent to most users than constructing a subclass, though that option would also still exist.

### Advantages
- Uniform interface for people searching for files of any type
- Convenience features of file class still available through return type
- Easier for scientists unfamiliar with python classes to extend and modify

### Disadvantages
- Mixed object-oriented and functional approach
- Later down selection of filelist type via array comprehension would require another call to maven_iuvs.filelist
  - Could maybe deal with this via a convenience downselect method that takes a function or list of bools and returns the downselected list

What am I missing? Let me know what you think. At the end of this PR I'd like a single set of file finders in the library, however we decide to proceed.

planetarymike commented:
It occurred to me that as a last resort we could just have two parallel file finding routines, but I think we can come to a solution that everyone is happy with.

One thing I was wondering about is if we're applying a class at the right level of abstraction. What if we had a class for an individual IUVS file rather than a list of files? Many of your smaller functions would make sense as member functions at that level, and that would also be the appropriate level to store some other file properties, instead of at the group or list level. Worth thinking about...

kconnour commented:
I thought about our discussion and I really think the fundamental difference is I was thinking that these classes could have methods that operate on filenames or data, whereas I think you said these should only operate on filenames / paths. I'd be happy if we did it this way. With this change, I would remove the methods I added related to relay files and reduce the number of classes that inherit each other, which I think are all things you'd like. I do think my original attempt made some good design choices that I'd like to keep, so I thought about what we both want for a while and propose the following:

First, I think we ought to have an "isolated" `IUVSDataFilenames` class. Given a string, it should be able to tell if it's an IUVS data filename and if it is, it should be able to pick out all info contained in the string (get level, get version number, get orbit number, etc.) I've previously written all these functions, so it'd take me very little time to combine them into a class. 

Next, I think we ought to have an `IUVSDataFiles` class that is the primary class for holding on to data filenames and their associated paths. This would likely be built heavily on `Pathlib` and just have basic methods to downselect files/paths based on a pattern, a boolean list, etc. I envision it taking a glob path and a glob pattern as inputs, and an optional index file. It would then glob for the pattern in path unless the index file is included, in which case it skips globbing and just uses that info. Then it'd make a list of `IUVSDataFilename` objects and remove the ones that aren't IUVS data files. This way we're sure an object of this class can only have IUVS data filenames. One of my favorite design choices of my previous attempt was allowing for a generic glob pattern and not helping the user in any way. My rationale is that sometimes I may want to use \*apoapse\*4000\*muv\* but Zac works with data across the whole mission so he may want to just glob 100 orbits at a time. He also works with both apoapse and periapse data an may want to get both of those at the same time for his analysis. I don't see a reasonable way to help the user in way that someone won't like (i.e. if we implemented *my* preference then Zac would have to make 100 of these objects and combine them. If we did it in a way better for Zac, then I have to get rid of a bunch of files). By leaving a generic glob pattern as an input, it doesn't limit the user's creativity.

For this reason, I propose we make a glob pattern creator class (which, I've largely done with the loose collection of functions at the bottom of `files.py`). This class would have a bunch of methods that help the user create patterns tailored to IUVS data, which can then be fed into `IUVSDataFiles`. I see this as the best of both worlds: there's a class that helps the user create a whatever pattern(s) suit their needs, and it doesn't limit the functionality of the main file holder class. 

So in summary, I'm proposing we have one class that extracts info from filenames, one class that creates glob patterns, and one class that essentially holds filenames and paths, along with methods that exclusively act on the list of filenames / paths. Thoughts on this proposal?

planetarymike commented:
Great! I think we're getting somewhere. I agree that there's a lot of good code in this PR and I'd like to find a way to incorporate a lot of it in whatever we end up with.

Sounds to me like you're suggesting three classes / code units:

1. `IUVSDataFilenames`, which holds a single filename and has some convenience methods for accessing members of the filename
2. `IUVSDataFiles`, which is a container for an iterable of `IUVSDataFilenames`. This has search methods that return a list of filenames on the system path.
3. A third class that helps to create glob patterns for a search pattern that is used by `IUVSDataFiles`

I think these are the right code units!

For the individual file container, what do you think about making this an IUVSDataFile object, with a class member that contains the HDUList object so that the user could do something like `myfile.fits.data['Primary']` to access elements of the FITS object? Users could potentially inherit from this if they are inclined to create e.g. an IUVSPeriapseFile object that has methods specific to the periapse data. I'd still have concerns about inheriting from this class, but file operations would be organized as a logical unit associated with a file object.

For the search, I agree that we need to provide an option to pass just a glob string in case the user wants to do something weird. But it's also nice to have options to load based on the filename tags b/c it's sometimes a pain to remember the order of the tags. I was thinking we might have a function like 

`def find_files(path=None, pattern=None, level=None, segment=None, orbit=None, channel=None, time=None):`

If the user passes something in pattern, the rest of the keys would be ignored, but if pattern=None, then we would call out to `make_iuvs_filename_glob` and construct the relevant glob string with the right orbit folders identified. the file finder would then perform the glob search, either against the index or against the file system. What do you think?

I think one way forward here would be for me to combine Zac's and my file finder functions with the features I envision, and then we can talk about the cost/benefit of attaching that function to a specific object. How does that sound?

kconnour commented:
> Great! I think we're getting somewhere. I agree that there's a lot of good code in this PR and I'd like to find a way to incorporate a lot of it in whatever we end up with.
> 
> Sounds to me like you're suggesting three classes / code units:
> 
> 1. `IUVSDataFilenames`, which holds a single filename and has some convenience methods for accessing members of the filename
> 2. `IUVSDataFiles`, which is a container for an iterable of `IUVSDataFilenames`. This has search methods that return a list of filenames on the system path.
> 3. A third class that helps to create glob patterns for a search pattern that is used by `IUVSDataFiles`
> 
> I think these are the right code units!

Yep, that's a good summary. Glad there's a solution we both like!

> 
> For the individual file container, what do you think about making this an IUVSDataFile object, with a class member that contains the HDUList object so that the user could do something like `myfile.fits.data['Primary']` to access elements of the FITS object? Users could potentially inherit from this if they are inclined to create e.g. an IUVSPeriapseFile object that has methods specific to the periapse data. I'd still have concerns about inheriting from this class, but file operations would be organized as a logical unit associated with a file object.

I would lean towards not implementing that functionality here, just because in my mind that's a lot closer to dealing with the file contents than the filenames / paths. I envision a rich set of classes for dealing with file contents where it seems more natural to me for such functionality to live, though that's well outside the scope of this pull request. I'd be ok with adding this though, and we could always move it at a later point if we desire.

> 
> For the search, I agree that we need to provide an option to pass just a glob string in case the user wants to do something weird. But it's also nice to have options to load based on the filename tags b/c it's sometimes a pain to remember the order of the tags. I was thinking we might have a function like
> 
> `def find_files(path=None, pattern=None, level=None, segment=None, orbit=None, channel=None, time=None):`
> 
> If the user passes something in pattern, the rest of the keys would be ignored, but if pattern=None, then we would call out to `make_iuvs_filename_glob` and construct the relevant glob string with the right orbit folders identified. the file finder would then perform the glob search, either against the index or against the file system. What do you think?

I certainly like the general idea here. My idea was that we could have functions that create glob strings from the class that helps create glob patterns (so they don't have to interact with that class directly), feeds these patterns into `IUVSDataFiles`, and returns an object of that type. Your proposal sounds pretty similar, except, if I understand you correctly, you're proposing one mega function that can create any search pattern instead of many smaller functions that can create specific search patterns. My concern with a function with 7 optional inputs is that the user would have to memorize that if `xxx` is included, the other inputs are ignored but if `xxx` isn't included, then certain other inputs are used. To me, it'd be simpler to have multiple functions that just do a single pattern creation. It wouldn't be a one stop shop for finding whatever files the user desires, but I think it'd be more maintainable. In any case, I see no reason why we couldn't implement both of our visions and let the user pick whichever one suits their needs best! 

> I think one way forward here would be for me to combine Zac's and my file finder functions with the features I envision, and then we can talk about the cost/benefit of attaching that function to a specific object. How does that sound?

That sounds good to me. It'd be beneficial for me to see specifically what you've got in mind. I'll push some updates later this weekend or early next week which, if you like them, may be beneficial when you're creating this function.

planetarymike commented:
I pushed a commit to this branch that replaces Zac's and my file finders with a single function that has all of their features. It has find_files and get_filename_glob_string, which implements most of the functionality of code units 2 and 3 from the list above.

planetarymike commented:
The more I think about it, the more I think we should have a fundamental IUVSFITS object that subclasses HDUList. We'd inherit all of the functionality of HDUList and keep all of the convenience features in your IUVSDataFilename. 

The rationale is that I expect the usual workflow is going to be to go directly from a find_files to fits.open() on the returned list of filenames, and then go on to performing operations on the FITS files. Returning a list of opened files that can be directly handled by the user would be a convenience. We can add a flag to return only the filenames and default to this behavior if the number of file objects that would be returned is really large (>1000 maybe?).

What do you think?

kconnour commented:
That sounds fine to me. Since it involves data I think it's getting a little bit away from the spirit of this pull request and can perhaps be its own request, but I'm fine with it either way.

planetarymike commented:
I just pushed a commit that adds an `IUVSFITS` class, a subclass of astropy's `HDUList`.

- All methods of `HDUList` are transparently wrapped
- Most methods / functions from `IUVSDataFilename` are incorporated with some streamlining:
  - check for IUVS Filename streamlined as glob match
  - time methods such as year delegated to `datetime`, e.g. `IUVSFITS.timestamp.year`
- `find_files` now returns a list of IUVSFITS objects, with other code updated to reflect this.

I didn't remove `IUVSDataFilename` with this commit, as that would entail breaking changes in `IUVSDataFilenames`. Keeping this class that operates on multiple files makes sense for e.g. grouping apoapsis files and performing common operations on them. I'd like to replace IUVSDataFilenames with IUVSDataFiles but there are a lot of interacting methods in `IUVSDataFilenames` that made it difficult for me to do this. Maybe you can take a crack at it Kyle?

At present there's a lot of duplicated search code between `IUVSDataFilenames` and the functions in search.py:

- `OrbitBlock`, `DataPath`, and `PatternGlob` overlap with `get_filename_glob_string`
- `GlobFiles` overlaps with the glob performed in `find_files`
- Some of the code in `IUVSDataFilenames` overlaps with `get_latest_files`

It would be nice if we can find a way to reduce this redundancy. My suggestion would be to use the freestanding functions inside `IUVSDataFilenames`, since these functions are fairly speedy, encapsulate basic functionality, and are unlikely to change, but I'm open to other alternatives. What do you think?

kconnour commented:
For the next few weeks I really need to focus on my radiative transfer code again, so it'll likely be a while before I get to investigate these pull requests.

planetarymike commented:
Hopefully we can work to merge some of this stuff in between some of our many other responsibilities, but I understand the pressure to get results on other projects. Let me know what I can do to make things easier--- in the meantime I'll keep pushing stuff :)

kconnour commented:
I've been hard at work on radiative transfer but I had a little time to look at these yesterday. 

I won't beat around the bush: I think we're suffering from a lack of organization / single responsibility. A number of functions in search.py don't seem to belong there (like `relay_file`), yet they're there because some higher level functions use them. In this case, the function is `get_apoapse_files` and I'm afraid it has waaay more than one responsibility---it finds files, but also performs checks of the mirror direction, determines the swath number of the file, etc. To your point in the previous message, a number of classes / functions use the same logic and I assert it's because most of them have more than one responsibility and the overlapping responsibility is where the duplicate logic comes into play. For this reason, I would actually propose that we delete almost all of these functions... though if you don't want to do something so drastic we could move them elsewhere. 

I thought about the responsibilities of a file finder and came up with 5:
* Create the path where data are located
* Create the glob search pattern
* Pull info from a filename
* Hold a collection of filenames, checking that the user's only using the most recent data
* Search for IUVS data files on one's computer

My idea is that items 1, 2, and 5 would be "standard" classes---classes that basically just perform operations---whereas 3 and 4 would be "data" classes---classes that only exist to extract and hold data from an input. And also, the methods in 5 would return an object of type 4. Before proceeding further, what do you think of this?

planetarymike commented:
Which functions specifically do you envision deleting? I want to encourage all users to commit code they find useful, so I'm not bothered by the multiple responsibilities of `get_apoapse_files`, or the presence of single-use functions. Once another user wanted to calculate swath numbers, it would make sense to split this out into separate code. If you're worried about clutter, we can start specifying `__all__` in the submodules so that only the most important routines are visible to the user. Or maybe your concern is that there should be a separate submodule for some of the code currently in `search.py`?

For file finding, I agree with your list. My preference is a more functional style, but I'm not opposed to objects where they make the code easier to understand.
- Create the path where data are located

    I think this is handled by some nested logic in `find_files` right now (under `if data_directory is None:`)--- we could split this out if you want. Otherwise the user specifies the path, which needs no special logic, right?
- Create the glob search pattern

    I implemented this as `get_filename_glob_string`--- I found a functional implementation more straightforward for this, since most people find imperative code easier to read and the inputs and outputs are very well defined. I'm not sure what an object wrapper adds?

- Pull info from a filename

    I agree that this is more properly the responsibility of an object with member functions. I implemented these as helper methods on top of `HDUList` in `IUVSFITS`. I like the idea of having an IUVS-specific wrapper around HDUList that users can inherit from for whatever they want to do. Seems to me that most of the additional info users might want to compute will be held at the file-level, rather than the file-list level, supporting encapsulation at this level.

- Hold a collection of filenames, checking that the user's only using the most recent data

    This seems to me to be two responsibilities. I implemented the recency check as a function `get_latest_files` that operates on an iterable of strings and returns only the most recent filenames in the list, eliminating old filenames. For the container of filenames or files, my initial hope is that most users will be able to get away with iterables of `IUVSFITS` or `HDUList` (or a derived class),  but I can see how a base class holding a list of files would be useful, with users that have a specific application able to inherit from this base.

- Search for IUVS data files on one's computer

  I implemented this as find_files, which is essentially a wrapper around the above four points. There is some parsing of the function options that's isn't strictly just a call to filesystem glob, but I find this logic to be pretty easy to understand, and feel that delegation to smaller functions would likely result in an overengineered and more confusing implementation.

To support both functional and object-oriented development, my preference would be to retain the functions as code units, and implement an `IUVSDataFiles` or `IUVSFITSList` class via calls to the functions. I didn't provide an example implementation because I'm not sure what all of the code in `IUVSDataFiles` is doing. I tried to understand the interaction of the methods in `IUVSDataFiles`, many of which operate on the object state by modifying object members, but I found it very confusing to have to hold the whole state of the object in my head in order to understand its basic behaviors. I'm not even sure what member variables the object has. Maybe you can help me out? What does / do you want this container to provide that's not available with a build-in iterable of HDUList or IUVSFITS?

kconnour commented:
> Which functions specifically do you envision deleting? I want to encourage all users to commit code they find useful, so I'm not bothered by the multiple responsibilities of `get_apoapse_files`, or the presence of single-use functions. Once another user wanted to calculate swath numbers, it would make sense to split this out into separate code. If you're worried about clutter, we can start specifying `__all__` in the submodules so that only the most important routines are visible to the user. Or maybe your concern is that there should be a separate submodule for some of the code currently in `search.py`?

This is largely in response to your request to reduce code duplication. Perhaps some responses below will help clarify what I'm suggesting.

> 
> For file finding, I agree with your list. My preference is a more functional style, but I'm not opposed to objects where they make the code easier to understand.
> 
>     * Create the path where data are located
>       I think this is handled by some nested logic in `find_files` right now (under `if data_directory is None:`)--- we could split this out if you want. Otherwise the user specifies the path, which needs no special logic, right?

This is perhaps my first bit of confusion. This is the responsibility of my `DataPattern` class. If we avoid duplication should I delete the class or try to integrate it with the logic in `find_files`? If we want this to be as general as possible it makes most sense to me to have this as a class. It seems to me that `find_files` assumes a user will have VM access and that's not necessarily the case... after all, if a user got data from PDS, data would be organized into Earth months instead of orbit blocks.

>     * Create the glob search pattern
>       I implemented this as `get_filename_glob_string`--- I found a functional implementation more straightforward for this, since most people find imperative code easier to read and the inputs and outputs are very well defined. I'm not sure what an object wrapper adds?

I have the same thoughts as before... are you requesting I remove my class or integrate my class into existing code? A class also makes the most sense to me here, where there's a fundamental method that's awfully similar to the function you're referring to. But most user's won't care to sort by timestamp, so the class could house methods related to making search patterns more tailored to their applications. 

> 
>     * Pull info from a filename
>       I agree that this is more properly the responsibility of an object with member functions. I implemented these as helper methods on top of `HDUList` in `IUVSFITS`. I like the idea of having an IUVS-specific wrapper around HDUList that users can inherit from for whatever they want to do. Seems to me that most of the additional info users might want to compute will be held at the file-level, rather than the file-list level, supporting encapsulation at this level.
> 
>     * Hold a collection of filenames, checking that the user's only using the most recent data
>       This seems to me to be two responsibilities. I implemented the recency check as a function `get_latest_files` that operates on an iterable of strings and returns only the most recent filenames in the list, eliminating old filenames. For the container of filenames or files, my initial hope is that most users will be able to get away with iterables of `IUVSFITS` or `HDUList` (or a derived class),  but I can see how a base class holding a list of files would be useful, with users that have a specific application able to inherit from this base.

> 
>     * Search for IUVS data files on one's computer
>       I implemented this as find_files, which is essentially a wrapper around the above four points. There is some parsing of the function options that's isn't strictly just a call to filesystem glob, but I find this logic to be pretty easy to understand, and feel that delegation to smaller functions would likely result in an overengineered and more confusing implementation.

I think the parsing is my concern. I'd personally prefer to figure out which smaller functions to use, then keep track of how a function with many inputs behaves if certain inputs are used.

> 
> 
> To support both functional and object-oriented development, my preference would be to retain the functions as code units, and implement an `IUVSDataFiles` or `IUVSFITSList` class via calls to the functions. I didn't provide an example implementation because I'm not sure what all of the code in `IUVSDataFiles` is doing. I tried to understand the interaction of the methods in `IUVSDataFiles`, many of which operate on the object state by modifying object members, but I found it very confusing to have to hold the whole state of the object in my head in order to understand its basic behaviors. I'm not even sure what member variables the object has. Maybe you can help me out? What does / do you want this container to provide that's not available with a build-in iterable of HDUList or IUVSFITS?

The idea is that you give this class a list of absolute paths of filenames and it takes advantage of the logic I wrote in `IUVSDataFilename`. It throws out any inputs that aren't IUVS data files, and uses the logic in that class to pick out info from the filename. Then it throws out all of the files which aren't the most recent files. It keeps the absolute path of the remaining "good" files, along with a corresponding list of `IUVSDataFilename` objects. 

It seems like we're moving in opposite directions due to philosophical differences. I'm not inclined to just give up, but it does seem like reconciling our different viewpoints will lead to obtuse solutions. Perhaps I can push a purely class-based implementation and users could select the one that works best for them?

planetarymike commented:
IUVSFITSList
------------------
In the interest of being concrete I pushed a commit that implements an IUVSFITSList class. I think this class captures all the user-facing methods of your IUVSDataFiles. Here are some of the choices I made in implementing this:

- The members are IUVSFITS objects rather than IUVSDataFilename objects, because I expect most users will want to interact with IUVS FITS immediately after doing a search. One less interface / piece of code to remember and write every time.
- Inherits from list so that it behaves like a list with some additional helper methods. Probably most users will want a thing called a list to have the basic attributes of a list, and this makes that easy.
- Implements `filenames`, `basenames` (names changed to conform to standard python) and your `downselect_` methods. I reduced the redundancy here by having the downselect_ methods all route to downselect_boolean behind the scenes. Downselects also return an IUVSFITSList of the matching files so there's no unexpected change of type.
- Definition of the `find_files` class method, which will enable users to populate this object with files found by find_files(). The arguments list and the docstring here is just a clone of find_files.
- - No checks on input files for the default initializer--- these are performed by `find_files`.

On its own IUVSFITSList is not that spectacular, but I can see how it might be very useful to those who want to implement an `IUVSApoapse` class or some equivalent that performs additional processing on a set of single-orbit apoapse files.

Generic Glob Matching
-------------------------------
I also modified the behavior of `find_files()` and `get_filename_glob_string` to be agnostic to the directory structure, to respond to your comment about making this generic. This comes with a major performance hit--- in the case of finding files from a specific orbit, the runtime is increased from 0.3 to 3s on my machine, a very noticeable difference. I'd rather just write some code that downloads data from the PDS into a directory structure matching that on the VM, so we can preserve a check for orbit directories--- I have this mostly implemented already.

Example use case
------------------------
I think one of the barriers in communication we have is regarding the typical use case of these objects. I couldn't find an example showing how to work with your file objects, so I was left wondering how everything fits together and what the purpose of the various pieces is. With the functions in search.py and the IUVSFITSList object, users could either call

> `maven_iuvs.search.find_files(segment='*corona', orbit=8204, channel='fuv')`

which would return a standard list of IUVSFITS objects, or 

> `iuvs.files.IUVSFITSList.find_files(segment='*corona', orbit=8204, channel='fuv')`

which would perform the same search but return an IUVSFITSList, with the extra helper methods you designed. 

What does a use case for your `IUVSDataFiles` look like?

Path forward
-----------------
I'd really like to get to a place where we agree on a fundamental set of file finding routines that can be a foundation for more specific work by individual users. I don't think having parallel implementations of something so important is a good idea. Separate interfaces to the same base code is a good idea though in order to support both OO and functional development styles. That's what I've tried to capture here--- objects that call functions from external packages and subpackages to encapsulate some of that behavior. Going the opposite way, with a functional wrapper around an object, would also be possible. If we go that route there are some details to work out:

1. All behaviors of find_files need to be available, since either Zac or I depend on them. This includes automatic detection of the data_directory, use of a file index, and the count argument.
2. `get_latest_files` is currently used not only for `find_files`, but also for syncing only the most recent data from the VM in `maven_iuvs.download.sync_data`. If the function is deleted the same behavior would need to be available to `sync_data`.

I propose that objects wrapping the existing functions is the path of least resistance. Unless there is something major missing from the current implementation of IUVSFITS and IUVSFITSList that would need to be present for your codebase to work?

kconnour commented:
I'd be ok with that proposal. I don't think there are any major features that I'd be missing if we just went with IUVSFITS and IUVSFITSLIST... and if there are I could always just re-add them. Honestly, IUVSFITS covers a good deal of the functionality of IUVSDataFiles so I'm fine if we just delete that class. Hopefully this will streamline the consolidation.

I'm really fighting with surface parameterization in FORTRAN for my RT code so I can approve changes but I'm still kinda lacking the time to make code contributions at the moment. Therefore, feel free to make as many of the aforementioned changes as you'd like and I'll be back whenever I can...

planetarymike commented:
Cool, so we can have some clean discussion on this I opened a new PR, #37. Closing this for now.
